File: app/static/js/super_admin.js
let currentOrgs = [];

document.addEventListener('DOMContentLoaded', () => {
    loadOverview();
    loadOrganizations();
});

async function loadOverview() {
    /**
     * Fetch aggregated counts for dashboard overview and populate the summary cards.
     * Side effects: updates DOM elements #totalOrgs, #totalAdmins, #totalStaff, #totalServices.
     * @returns {Promise<void>}
     */
    try {
        const response = await fetch('/super-admin/api/overview');
        const data = await response.json();

        document.getElementById('totalOrgs').textContent = data.total_organizations;
        document.getElementById('totalAdmins').textContent = data.total_admins;
        document.getElementById('totalStaff').textContent = data.total_staff;
        document.getElementById('totalServices').textContent = data.total_services;
    } catch (error) {
        console.error('Error loading overview:', error);
    }
}

function showTab(tabName) {
    // Switch visible tab: remove active state from all and hide their content
    document.querySelectorAll('.tab').forEach(tab => tab.classList.remove('active'));
    document.querySelectorAll('.tab-content').forEach(content => content.style.display = 'none');

    // `event` is expected to be the click event from the element that triggered this function.
    // Caller must ensure `showTab` is invoked from an event handler (e.g. onclick="showTab('admins')").
    if (event && event.target) event.target.classList.add('active');
    document.getElementById(`${tabName}Tab`).style.display = 'block';

    // Load additional data when opening specific tabs
    if (tabName === 'admins') {
        loadAdmins();
    }
}

// Organizations Management
async function loadOrganizations() {
    /**
     * Load organizations for management table and cache them in `currentOrgs`.
     * Side effects: sets `currentOrgs` and renders rows into #orgsBody.
     * @returns {Promise<void>}
     */
    try {
        const response = await fetch('/super-admin/api/organizations');
        currentOrgs = await response.json();

        const tbody = document.getElementById('orgsBody');
        tbody.innerHTML = currentOrgs.map(org => `
            <tr>
                <td><strong>${org.name}</strong></td>
                <td>${org.location || '-'}</td>
                <td>${org.contact || '-'}</td>
                <td>${org.admin_count}</td>
                <td>${org.service_count}</td>
                <td>${org.staff_count}</td>
                <td>
                    <button onclick="editOrganization(${org.id})" class="btn btn-primary">Edit</button>
                    <button onclick="deleteOrganization(${org.id})" class="btn btn-danger">Delete</button>
                </td>
            </tr>
        `).join('');
    } catch (error) {
        console.error('Error loading organizations:', error);
    }
}

function showOrgForm() {
    /**
     * Show the organization add form and clear inputs for a fresh create.
     * Side effects: manipulates #orgForm inputs and title.
     */
    document.getElementById('orgForm').style.display = 'block';
    document.getElementById('orgFormTitle').textContent = 'Add Organization';
    document.getElementById('orgId').value = '';
    document.getElementById('orgName').value = '';
    document.getElementById('orgLocation').value = '';
    document.getElementById('orgContact').value = '';
}

function editOrganization(id) {
    /**
     * Populate the organization form with existing org values for editing.
     * @param {number} id - organization id to edit
     */
    const org = currentOrgs.find(o => o.id === id);
    if (org) {
        document.getElementById('orgForm').style.display = 'block';
        document.getElementById('orgFormTitle').textContent = 'Edit Organization';
        document.getElementById('orgId').value = org.id;
        document.getElementById('orgName').value = org.name;
        document.getElementById('orgLocation').value = org.location || '';
        document.getElementById('orgContact').value = org.contact || '';
    }
}

async function saveOrganization() {
    /**
     * Create or update an organization based on whether #orgId has a value.
     * Sends JSON to server and refreshes the table and overview on success.
     * @returns {Promise<void>}
     */
    const id = document.getElementById('orgId').value;
    const data = {
        name: document.getElementById('orgName').value,
        location: document.getElementById('orgLocation').value,
        contact: document.getElementById('orgContact').value
    };

    try {
        const url = id ? `/super-admin/api/organizations/${id}` : '/super-admin/api/organizations';
        const method = id ? 'PUT' : 'POST';

        const response = await fetch(url, {
            method: method,
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data)
        });

        if (response.ok) {
            cancelOrgForm();
            loadOrganizations();
            loadOverview();
        }
    } catch (error) {
        console.error('Error saving organization:', error);
    }
}

async function deleteOrganization(id) {
    /**
     * Delete an organization by id after user confirmation.
     * Side effects: removes org server-side, then refreshes list and overview.
     * @param {number} id - id of organization to delete
     * @returns {Promise<void>}
     */
    if (!confirm('Delete this organization? This will also delete all related services and staff.')) return;

    try {
        const response = await fetch(`/super-admin/api/organizations/${id}`, { method: 'DELETE' });
        if (response.ok) {
            loadOrganizations();
            loadOverview();
        }
    } catch (error) {
        console.error('Error deleting organization:', error);
    }
}

function cancelOrgForm() {
    /**
     * Hide the organization form without saving changes.
     */
    document.getElementById('orgForm').style.display = 'none';
}

// Admins Management
async function loadAdmins() {
    /**
     * Load admin users and render them into the admins table.
     * Side effects: updates DOM in #adminsBody.
     * @returns {Promise<void>}
     */
    try {
        const response = await fetch('/super-admin/api/admins');
        const admins = await response.json();

        const tbody = document.getElementById('adminsBody');
        tbody.innerHTML = admins.map(admin => `
            <tr>
                <td><strong>${admin.username}</strong></td>
                <td>${admin.organization_name}</td>
                <td>
                    <button onclick="editAdmin(${admin.id})" class="btn btn-primary">Edit</button>
                    <button onclick="deleteAdmin(${admin.id})" class="btn btn-danger">Delete</button>
                </td>
            </tr>
        `).join('');
    } catch (error) {
        console.error('Error loading admins:', error);
    }
}

function showAdminForm() {
    /**
     * Display admin creation form and populate organization dropdown from cached `currentOrgs`.
     */
    document.getElementById('adminForm').style.display = 'block';
    document.getElementById('adminId').value = '';
    document.getElementById('adminUsername').value = '';
    document.getElementById('adminPassword').value = '';

    const select = document.getElementById('adminOrg');
    select.innerHTML = '<option value="">Select Organization</option>' +
        currentOrgs.map(o => `<option value="${o.id}">${o.name}</option>`).join('');
}

function editAdmin(id) {
    showAdminForm();
    document.getElementById('adminId').value = id;
    document.getElementById('adminUsername').disabled = true;
    document.getElementById('adminPassword').placeholder = 'Leave empty to keep current';
}

async function saveAdmin() {
    /**
     * Create or update an admin account. If updating, keep username disabled.
     * Sends JSON to server and refreshes admins and overview after success.
     * @returns {Promise<void>}
     */
    const id = document.getElementById('adminId').value;
    const data = {
        username: document.getElementById('adminUsername').value,
        password: document.getElementById('adminPassword').value,
        organization_id: document.getElementById('adminOrg').value || null
    };

    try {
        const url = id ? `/super-admin/api/admins/${id}` : '/super-admin/api/admins';
        const method = id ? 'PUT' : 'POST';

        const response = await fetch(url, {
            method: method,
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data)
        });

        if (response.ok) {
            cancelAdminForm();
            loadAdmins();
            loadOverview();
        } else {
            const error = await response.json();
            alert(error.error || 'Failed to save admin');
        }
    } catch (error) {
        console.error('Error saving admin:', error);
    }
}

async function deleteAdmin(id) {
    /**
     * Delete an admin account after confirmation and refresh lists.
     * @param {number} id - admin id to delete
     * @returns {Promise<void>}
     */
    if (!confirm('Delete this admin?')) return;

    try {
        const response = await fetch(`/super-admin/api/admins/${id}`, { method: 'DELETE' });
        if (response.ok) {
            loadAdmins();
            loadOverview();
        }
    } catch (error) {
        console.error('Error deleting admin:', error);
    }
}

function cancelAdminForm() {
    /**
     * Hide the admin form and re-enable username input.
     */
    document.getElementById('adminForm').style.display = 'none';
    document.getElementById('adminUsername').disabled = false;
}

File: app/routes/super_admin.py
from flask import Blueprint, render_template, request, jsonify, session, redirect, url_for
from app.models import db, User, Organization, Service
from functools import wraps

bp = Blueprint('super_admin', __name__, url_prefix='/super-admin')

def super_admin_required(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if 'user_id' not in session or session.get('role') != 'super_admin':
            return redirect(url_for('super_admin.login'))
        return f(*args, **kwargs)
    return decorated_function

@bp.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'GET':
        return render_template('super_admin_login.html')
    
    data = request.json
    username = data.get('username')
    password = data.get('password')
    
    user = User.query.filter_by(username=username, role='super_admin').first()
    if user and user.check_password(password):
        session['user_id'] = user.id
        session['username'] = user.username
        session['role'] = user.role
        return jsonify({'success': True})
    
    return jsonify({'error': 'Invalid credentials'}), 401

@bp.route('/logout')
def logout():
    session.clear()
    return redirect(url_for('super_admin.login'))

@bp.route('/dashboard')
@super_admin_required
def dashboard():
    return render_template('super_admin_dashboard.html')

# Organization Management
@bp.route('/api/organizations', methods=['GET'])
@super_admin_required
def get_organizations():
    """Get all organizations with stats"""
    orgs = Organization.query.all()
    result = []
    
    for org in orgs:
        admin_count = User.query.filter_by(organization_id=org.id, role='admin').count()
        service_count = Service.query.filter_by(organization_id=org.id).count()
        staff_count = User.query.filter_by(organization_id=org.id, role='staff').count()
        
        result.append({
            **org.to_dict(),
            'admin_count': admin_count,
            'service_count': service_count,
            'staff_count': staff_count
        })
    
    return jsonify(result)

@bp.route('/api/organizations', methods=['POST'])
@super_admin_required
def create_organization():
    """Create new organization"""
    data = request.json
    
    org = Organization(
        name=data['name'],
        location=data.get('location', ''),
        contact=data.get('contact', '')
    )
    db.session.add(org)
    db.session.commit()
    return jsonify(org.to_dict())

@bp.route('/api/organizations/<int:org_id>', methods=['PUT'])
@super_admin_required
def update_organization(org_id):
    """Update organization"""
    org = Organization.query.get(org_id)
    if not org:
        return jsonify({'error': 'Organization not found'}), 404
    
    data = request.json
    org.name = data.get('name', org.name)
    org.location = data.get('location', org.location)
    org.contact = data.get('contact', org.contact)
    
    db.session.commit()
    return jsonify(org.to_dict())

@bp.route('/api/organizations/<int:org_id>', methods=['DELETE'])
@super_admin_required
def delete_organization(org_id):
    """Delete organization"""
    org = Organization.query.get(org_id)
    if not org:
        return jsonify({'error': 'Organization not found'}), 404
    
    db.session.delete(org)
    db.session.commit()
    return jsonify({'success': True})

# Admin Management
@bp.route('/api/admins', methods=['GET'])
@super_admin_required
def get_admins():
    """Get all organization admins"""
    admins = User.query.filter_by(role='admin').all()
    result = []
    
    for admin in admins:
        org = Organization.query.get(admin.organization_id) if admin.organization_id else None
        result.append({
            **admin.to_dict(),
            'organization_name': org.name if org else 'None'
        })
    
    return jsonify(result)

@bp.route('/api/admins', methods=['POST'])
@super_admin_required
def create_admin():
    """Create new organization admin"""
    data = request.json
    
    # Check if username exists
    if User.query.filter_by(username=data['username']).first():
        return jsonify({'error': 'Username already exists'}), 400
    
    admin = User(
        username=data['username'],
        role='admin',
        organization_id=data.get('organization_id')
    )
    admin.set_password(data['password'])
    
    db.session.add(admin)
    db.session.commit()
    return jsonify(admin.to_dict())

@bp.route('/api/admins/<int:admin_id>', methods=['PUT'])
@super_admin_required
def update_admin(admin_id):
    """Update admin"""
    admin = User.query.filter_by(id=admin_id, role='admin').first()
    if not admin:
        return jsonify({'error': 'Admin not found'}), 404
    
    data = request.json
    admin.organization_id = data.get('organization_id', admin.organization_id)
    
    if data.get('password'):
        admin.set_password(data['password'])
    
    db.session.commit()
    return jsonify(admin.to_dict())

@bp.route('/api/admins/<int:admin_id>', methods=['DELETE'])
@super_admin_required
def delete_admin(admin_id):
    """Delete admin"""
    admin = User.query.filter_by(id=admin_id, role='admin').first()
    if not admin:
        return jsonify({'error': 'Admin not found'}), 404
    
    db.session.delete(admin)
    db.session.commit()
    return jsonify({'success': True})

# Overview Stats
@bp.route('/api/overview', methods=['GET'])
@super_admin_required
def overview():
    """Get system overview statistics"""
    total_orgs = Organization.query.count()
    total_admins = User.query.filter_by(role='admin').count()
    total_staff = User.query.filter_by(role='staff').count()
    total_services = Service.query.count()
    
    return jsonify({
        'total_organizations': total_orgs,
        'total_admins': total_admins,
        'total_staff': total_staff,
        'total_services': total_services
    })